core/drivers/zynqmp_pm.c

```diff
@@ -0,0 +1,115 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2021 Foundries.io Ltd
+ */
+
+#include <arm.h>
+#include <drivers/zynqmp_pm.h>
+#include <kernel/cache_helpers.h>
+#include <kernel/thread.h>
+#include <mm/core_memprot.h>
+#include <string.h>
+#include <tee/cache.h>
+#include <tee_api_types.h>
+#include <types_ext.h>
+#include <utee_defines.h>
+
+/*
+ * For additional details about ZynqMP specific SMC ID's and PM request
+ * handling in TF-A check
+ * https://xilinx-wiki.atlassian.net/wiki/spaces/A/pages/18842107/Arm+Trusted+Firmware
+ */
+#define EFUSE_ACCESS_SMC	0xC2000035
+#define EFUSE_NOT_ENABLED	29
+
+enum efuse_op {EFUSE_READ = 0, EFUSE_WRITE = 1};
+
+/*
+ * Stores all required details to
+ * read/write efuse memory.
+ * @src:        paddr_t of the buffer to store the data to be write/read
+ *              buffer should be cache aligned
+ * @size:       number of 32-bit words to be read/written
+ * @offset:     offset to be read/written
+ * @flag:       EFUSE_READ  - represents efuse read operation
+ *              EFUSE_WRITE - represents efuse write operation
+ * @pufuserfuse:0 - represents non-puf efuses, offset is used for read/write
+ *              1 - represents puf user fuse row number.
+ */
+struct xilinx_efuse {
+	uint64_t src;
+	uint32_t size;
+	uint32_t offset;
+	uint32_t flag;
+	uint32_t pufuserfuse;
+};
+
+static uint32_t zynqmp_sip_call(uint32_t pm_api_id, uint32_t arg0,
+				uint32_t arg1, uint32_t arg2, uint32_t arg3,
+				uint32_t *payload)
+{
+	struct thread_smc_args args = {
+		.a0 = pm_api_id,
+		.a1 = reg_pair_to_64(arg1, arg0),
+		.a2 = reg_pair_to_64(arg3, arg2),
+	};
+
+	thread_smccc(&args);
+
+	if (payload)
+		*payload = args.a0 >> 32;
+
+	return args.a0;
+}
+
+static TEE_Result efuse_op(enum efuse_op op, uint8_t *buf, size_t sz,
+			   uint32_t efuse_offset, bool puf)
+{
+	paddr_t addr = 0;
+	uint32_t res = 0;
+	TEE_Result ret = TEE_SUCCESS;
+	struct xilinx_efuse *efuse = NULL;
+
+	if (!buf)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	efuse = malloc(sizeof(*efuse));
+	if (!efuse)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	efuse->src = virt_to_phys(buf);
+	efuse->size = sz / sizeof(uint32_t);
+	efuse->offset = efuse_offset;
+	efuse->flag = op;
+	efuse->pufuserfuse = puf;
+
+	cache_operation(TEE_CACHECLEAN, (void *)buf, sz);
+	cache_operation(TEE_CACHECLEAN, (void *)efuse, sizeof(*efuse));
+
+	addr = virt_to_phys(efuse);
+
+	res = zynqmp_sip_call(EFUSE_ACCESS_SMC,
+			      (uint32_t)(addr >> 32),
+			      (uint32_t)(addr),
+			      0, 0, NULL);
+	if (res) {
+		if (res == EFUSE_NOT_ENABLED)
+			EMSG("Efuse access is not enabled");
+		else
+			EMSG("Error in efuse access %x", res);
+
+		ret = TEE_ERROR_GENERIC;
+	}
+
+	cache_operation(TEE_CACHEINVALIDATE, (void *)buf, sz);
```

Comment:
`buf` must be cacheline aligned and it's size a multiple of the cacheline size for this to work or you may invalidate unrelated data.
The cast isn't needed.

Reply:
what is the problem with flushing more data from the cache than necessary? is it really worth asking for aligned memory to avoid flushing a few extra bytes?

Reply:
We are invalidating here. Flush (clean and invalidate) can't be used here due to cache speculation.

Reply:
this is done after PMU writes data from caches, flushing can potentially overwrite this data in RAM written by PMU.
IMO having temporary buffers, like it was done from the very beginning would make our life much easier

Reply:
@jenswi-linaro @ldts 

To address this comment I see two options:
1) bring back that temporary buffers, that are allocated within the function with memalign and size a multiple of the cacheline size
2) Add a check is buf is aligned here for the buffer provided as param as function (so imposes a the caller of that function to allocate it correctly), but that will require additional param to pass to the function (besides buffer size we should also provides amount of bytes we want to read from eFUSE):

```
       cacheline_size = dcache_get_line_size();
       if (!IS_ALIGNED((uintptr_t)buf,cacheline_size) ||
           (sz % cacheline_size)) {
               EMSG("Buffer should be cache aligned");
               return TEE_ERROR_BAD_PARAMETERS;
       }

```

Reply:
@jenswi-linaro could you please provide your input about the preferred way to address this?  thanks

Reply:
Option 1 sounds good to me.

Reply:
Fixed